src/applications/bmqstoragetool/m_bmqstoragetool_searchresult.h

```diff
@@ -0,0 +1,630 @@
+
+    // MANIPULATORS
+
+    /// Process `message` record with the specified `record`, `recordIndex` and
+    /// `recordOffset`.
+    bool processMessageRecord(const mqbs::MessageRecord& record,
+                              bsls::Types::Uint64        recordIndex,
+                              bsls::Types::Uint64        recordOffset)
+        BSLS_KEYWORD_OVERRIDE;
+    /// Process `confirm` record with the specified `record`, `recordIndex` and
+    /// `recordOffset`.
+    bool processConfirmRecord(const mqbs::ConfirmRecord& record,
+                              bsls::Types::Uint64        recordIndex,
+                              bsls::Types::Uint64        recordOffset)
+        BSLS_KEYWORD_OVERRIDE;
+    /// Process `deletion` record with the specified `record`, `recordIndex`
+    /// and `recordOffset`.
+    bool processDeletionRecord(const mqbs::DeletionRecord& record,
+                               bsls::Types::Uint64         recordIndex,
+                               bsls::Types::Uint64         recordOffset)
+        BSLS_KEYWORD_OVERRIDE;
+    /// Output result of a search.
+    void outputResult() BSLS_KEYWORD_OVERRIDE;
+    /// Output result of a search filtered by the specified GUIDs filter.
+    void outputResult(const bsl::unordered_set<bmqt::MessageGUID>& guidFilter)
+        BSLS_KEYWORD_OVERRIDE;
+};
+
+// ========================
+// class SearchDetailResult
+// ========================
+
+/// This class provides logic to handle and output detail result.
+class SearchDetailResult : public SearchResult {
+  private:
+    // PRIVATE TYPES
+
+    typedef bsl::unordered_map<bmqt::MessageGUID, MessageDetails>
+        MessagesDetails;
+    // Type that represents map for guid and message details.
+
+    // PRIVATE DATA
+
+    bsl::ostream& d_ostream;
+    // Reference to output stream.
+    const QueueMap& d_queueMap;
+    // Reference to 'QueueMap' instance.
+    const bslma::ManagedPtr<PayloadDumper> d_payloadDumper;
+    // Pointer to 'PayloadDumper' instance.
+    const bool d_printImmediately;
+    // If true, print message details as soon as it is complete (usually when
+    // 'deleted' record received) to save memory. If false, message data
+    // remains stored in MessagesDetails for further processing.
+    const bool d_eraseDeleted;
+    // If true, erase data from MessagesDetails when 'deleted' record is
+    // received to save memory. If false, message data remains stored in
+    // MessagesDetails for further processing.
+    const bool d_cleanUnprinted;
+    // If true, clean remaining data in MessagesDetails before printing final
+    // result.
+    bsl::size_t d_printedMessagesCount;
+    // Printed messages count.
+    MessagesDetails d_messagesDetails;
+    // Storage for messages details.
+    bsl::map<bsls::Types::Uint64, bmqt::MessageGUID> d_messageIndexToGuidMap;
+    // Map to store sorted indexes to preserve messages order for output.
```

Comment:
Use list instead of map to retain messages order. We don't actually need to sort them so there is no reason to "pay" logarithmic time for each insertion and deletion.